### DISCUSSION TITLE
We need a way for the community to adopt abandoned repos

### DISCUSSION THREAD

**Discussion by endolith** [community/community#13470] [Other Feature Feedback, Questions, & Ideas]
Sometimes there's a useful repo, but the original author has abandoned it.  

Of course it's open source and decentralized, so anyone can continue to fork it and make changes, but it's *too* decentralized: The forks aren't very visible, which results in many people forking the repo to make the same change, fixing the same problem, unaware that there are already forks that do the same:

![88466224-1459c600-ce98-11ea-9994-dc31a23dfd36 1](https://user-images.githubusercontent.com/58611/160050584-2d8c5081-64dd-4e30-9dc2-be145e43e90e.png)

Development is fragmented and stalled, the original Issues stay unresolved forever, the original Pull Requests stay unmerged forever, commands like `pip install packagename` no longer include the latest developments, etc.  There's no great way to collaborate after a main repo is dead.

Some examples I know of:

* https://github.com/aaren/wavelets/issues/22 (Author has disappeared, many people have made the same change for compatibility)
* https://github.com/RedaOps/ann-visualizer/network (not touched in 4 years, but people are still making their own versions with the useful features left hanging on the original repo.)
* https://github.com/anazalea/pySankey/issues/41

There really should be some kind of adoption process where the community can continue to maintain things after the author has disappeared.

Related:
* https://github.com/isaacs/github/issues/1385
* [How should I contribute to a (mostly) abandoned GitHub project?](https://softwareengineering.stackexchange.com/q/273156/7525)
* [A suggestion addressing the ABANDONED REPOSITOREIS](https://github.community/t/a-suggestion-addressing-the-abandoned-repositoreis/144264)
* [What is GitHub policy about abandoned repositories?](https://github.community/t/what-is-github-policy-about-abandoned-repositories/117668)
* [Lovely forks](https://addons.mozilla.org/en-US/firefox/addon/lovely-forks/): Show notable forks of Github projects.
* https://github.com/github/feedback/discussions/5185
* https://github.com/github/feedback/discussions/12991

I don't know the best solution.  Just showing that there is an active fork, copying or shadowing Issues and Pull Requests to other forks?  Or converting it into a community project with new maintainers being elected by others who committed to the original repo or something?  But the current situation is not good.  

Maybe each answer can be a proposed solution.

**Comment by endolith** [comment]
I've been trying to consolidate the forks of the ann-visualizer project. I cannot comment on or otherwise change the pull requests in the archived repo, but I can link to them from the commits in my own fork and the PR then shows that I've referenced it.  I can also @ notify the original PR authors, and they can delete their own pull requests and forks.  Then I can merge forks that have been implemented in some way or other, so they show up as merged in the network graph.  It's pretty painful and you still wouldn't know there's an active fork by looking at the main repo, unless you look at the network graph, but it's a partial solution.

**Reply by Khanditanti7** [reply]
Please give me a few minutes and I will get back to you on this one.

**Comment by NitinIsFit** [comment]
How people are supposed to adopt the repo without author's permission? There should be an option like `"@user wants to adopt your read-only repo, are you willing to transfer them the ownership?, Be careful because this option cannot be undone"`.
And two buttons saying `"yes"` and `"decline request"`.
So author could chose if he want to transfer the ownership to other person who he finds appropriate to be a new owner, if he doesn't wanna transfer the ownership he will simply decline the offer.

**Reply by endolith** [reply]
This is more about cases where the author would never respond to such a request, positively or negatively, or even acknowledge it at all (because they are dead, or retired, or have disappeared to live off-the-grid forever, etc.)

**Comment by DeathByNukes** [comment]
Another example: the repository "dnSpy" was suddenly archived and abandoned with no explanation, and the community fork "dnSpyEx" is not easily discoverable unless you click the forks link on the right (not the big button, the small link) and find the one with the most stars.

I think archived projects ought to at least have a panel at the top dedicated to helping visitors find alternatives. In this example, simply listing the fork with the most stars would be helpful.

In this discussion https://github.com/dnSpyEx/dnSpy/discussions/76 I saw them considering un-forking, supposedly because GitHub makes forks hard to find in searches. If that is still true, there ought to be an exception for forks of archived or dead projects, and probably also an exception for forks with a lot of stars or commits. Goading a project into un-forking seems counterproductive because it lowers its apparent credibility and removes it from the list of forks.

Thank you.

**Comment by smheidrich** [comment]
I opened a related feature request here: [Display community-suggested alternatives for archived repos](https://github.com/orgs/community/discussions/158306)

The idea is that the community would be able to suggest and vote for alternatives to archived repositories. This wouldn't be tied to forks so that even "spiritual successors" that were written from scratch but fulfill the same purpose could be suggested.

It might look something like this **(mockup)**:

![UI mockup showing a community-suggested alternative being displayed in the archived-repo banner](https://github.com/user-attachments/assets/25379baa-82f1-4d5e-86b8-89bf63bf2633)

Haven't put any thought into how voting etc. would work in detail, but I'm sure it could be figured out.

**Comment by BasicHtmlCoder** [comment]
I don't think there is any Git solution to this. It is so relative as when to consider a valuable repository abandoned, as Git is trying to do things mathematically effective. Which leads to what you say too decentralized.

But let's try for the , I suggest such metric in pseudo SQL: 

```SQL
SELECT * FROM repos WHERE repo.stars > 100 AND (now() - repo.last_update) < 1 YEAR AS abandoned
SELECT forks FROM abandoned WHERE abandoned.stars < SOME(abandoned.forks.stars * 3) -- get repos with some forks having at least the 
```

half number of stars, so if a repo have 1000 stars, one of its forks has to have at least 334 stars to candidate as an alternative, Otherwise even if the fork has stars It would not qualify as a liable replacement.


This seems to be a good query, (thanks chatgpt)
```
WITH base_repos AS (
  SELECT
    repo_name,
    owner,
    name,
    watchers_count AS stars,
    pushed_at,
    fork,
    parent_name
  FROM `bigquery-public-data.github_repos.repos`
),

-- Step 1: popular repos
popular_repos AS (
  SELECT *
  FROM base_repos
  WHERE stars > 100
),

-- Step 2: abandoned repos (not updated in >5 years)
abandoned_repos AS (
  SELECT *
  FROM popular_repos
  WHERE pushed_at < TIMESTAMP_SUB(CURRENT_TIMESTAMP(), INTERVAL 5 YEAR)
     OR pushed_at IS NULL
),

-- Step 3: forks with their star counts
forks AS (
  SELECT
    repo_name AS fork_repo,
    owner     AS fork_owner,
    stars     AS fork_stars,
    parent_name
  FROM base_repos
  WHERE fork = TRUE
),

-- Step 4: join forks to abandoned parents
successful_forks AS (
  SELECT
    a.repo_name  AS original_repo,
    a.owner      AS original_owner,
    a.stars      AS original_stars,
    f.fork_repo,
    f.fork_owner,
    f.fork_stars
  FROM abandoned_repos a
  JOIN forks f
    ON f.parent_name = a.repo_name
  WHERE f.fork_stars > a.stars * 3
)

-- Final output
SELECT *
FROM successful_forks
ORDER BY fork_stars DESC
LIMIT 1000;
```

if anyone can execute it on BigQuery and provide the export. I would be thankful
